### DISCUSSION TITLE
Some Characteristics about libiiwa.

### DISCUSSION THREAD

**Discussion by CodingCatMountain** [Toni-SM/libiiwa#3] [Q&A]
Hello, developer! @Toni-SM . Thank you for this repo! There is a KUKA iiwa7 in my lab. So far，we utilize the ``iiwa_stack`` or even "iiwa_ros" repo to control KUKA iiwa7. However, we have met some problems. 1. iiwa_stack is too slow due to the Protocol used in this repo. And ``iiwa_stack`` does not provide the API to control joint_torque. So a lot of useful controller, e.g., LQR, could not be implemented via ``iiwa_stack``.  2. iiwa_ros is not very stable. For example, KUKA iiwa7 will randomly swing when the control mode is set as "Position". And there are some problems happen in gravity compensation when the tool is attached on the end-effector of iiwa7. So I wanna to ask some questions about this repo. Which protocol used in this repo? Does any controller example rely on the joint_torque have been implemented on the real robot via this repo? Or could you provide more real-robot examples which  the input of the controller running on the real robot is joint_torque? Looking forward to your reply! :-)

**Comment by Toni-SM** [comment]
Hi @CodingCatMountain 

Current `libiiwa` implementation relies on the KUKA Standard and Servoing interfaces:
- Standard interface: A software interface that allows viewing the current state of the robot, setting parameters, and programming discontinuous blocking or non-blocking motion sequences.
- Servo (or servoing) interface: A real-time software interface that allows the implementation of non-deterministic applications and continuous and smooth motion composed of multiple points. This interface requires the presence of some KUKA-specific software libraries in the cabinet (used by `iiwa_stack`).

Neither of them allows the application of torque for robot control. 

For this, it is necessary to rely on the KUKA Fast Robot Interface (FRI), a fast robot interface that enables access to the robot controller from an external computer in real time. FRI can only be controlled in C++ (KUKA Fast Robot Interface C++ SDK) and requires JAVA  to initialize the FRI interface at robot (cabinet) side. 

The idea is to eventually add FRI support to `libiiwa`... although I don't know for when

**Reply by CodingCatMountain** [reply]
@Toni-SM Thank you for your replay. Toni. OK, I understand. Hope someday KUKA iiwa could be easily used as Franka emika do. :-)
